src/models/room.ts

```diff
@@ -2249,6 +2288,184 @@ export class Room extends EventEmitter {
+        const powerLevelsEvents = this.currentState.getStateEvents(EventType.RoomPowerLevels, "");
+        const powerLevels = powerLevelsEvents && powerLevelsEvents.getContent();
+        if (!powerLevels || powerLevels["org.matrix.msc3531.visibility"] < event.sender.powerLevel) {
+            return;
+        }
+
+        // Record this change in visibility.
+        // If the event is not in our timeline and we only receive it later,
+        // we may need to apply the visibility change at a later date.
+
+        const visibilityChangesOnOriginalEvent = this.visibilityChanges.get(originalEventId);
+        if (visibilityChangesOnOriginalEvent) {
+            // It would be tempting to simply erase the latest visibility change
+            // but we need to record all of the changes in case the latest change
+            // is ever redacted.
+            // In practice, it should be unusual for an event to have more than
+            // 2 visibility changes, so linear scans through `visibilityChanges`
+            // should be fast. It might be possible to DoS this feature, though.
+            let index = visibilityChangesOnOriginalEvent.length - 1;
+            for (; index >= 0; --index) {
+                const target = visibilityChangesOnOriginalEvent[index];
+                if (target.getTs() < event.getTs()) {
+                    break;
+                }
+            }
+            if (index == -1) {
+                visibilityChangesOnOriginalEvent.unshift(event);
+            } else {
+                visibilityChangesOnOriginalEvent.splice(index + 1, 0, event);
+            }
+        } else {
+            this.visibilityChanges.set(originalEventId, [event]);
+        }
+
+        // Finally, let's check if the event is already in our timeline.
+        // If so, we need to patch it and inform listeners.
+
+        const originalEvent = this.findEventById(originalEventId);
+        if (!originalEvent) {
+            return;
+        }
+        if (originalEvent.applyVisibilityChange(visible, reason)) {
+            this.emit("Room.visibilityChange", event);
+        }
+    }
+
+    private redactVisibilityChangeEvent(event: MatrixEvent) {
+        // Sanity checks.
+        if (!event.isEventVisibilityChange) {
+            throw new Error("expected a visibility change event");
+        }
+        const relation = event.getRelation();
+        const originalEventId = relation.event_id;
+        const visibilityChangesOnOriginalEvent = this.visibilityChanges.get(originalEventId);
+        if (!visibilityChangesOnOriginalEvent) {
+            // No visibility changes on the original event.
+            // In particular, this change event was not recorded,
+            // most likely because it was ill-formed.
+            return;
+        }
+        const index = visibilityChangesOnOriginalEvent.findIndex(change => change.getId() == event.getId());
+        if (index == -1) {
```

Comment:
```suggestion
        if (index === -1) {
```

Reply:
this sort of comment applies throughout - going to stop posting about it for this review :)